core/src/main/scala/shapeless/lazy.scala

```diff
@@ -46,48 +47,187 @@ object Lazy {
   def values[T <: HList](implicit lv: Lazy[Values[T]]): T = lv.value.values
 
   implicit def mkLazy[I]: Lazy[I] = macro LazyMacros.mkLazyImpl[I]
+
+  trait Cached[+T] extends Lazy[T] {
+    override def map[U](f: T => U): Cached[U] = Cached { f(value) }
+    def flatMap[U](f: T => Cached[U]): Cached[U] = Cached { f(value).value }
+  }
+
+  object Cached {
+    implicit def apply[T](t: => T): Cached[T] =
+      new Cached[T] {
+        val value = t
+      }
+
+    def unapply[T](lt: Cached[T]): Option[T] = Some(lt.value)
+
+    implicit def mkCached[I]: Cached[I] = macro LazyMacros.mkCachedLazyImpl[I]
+  }
+
 }
 
 object lazily {
   def apply[T](implicit lv: Lazy[T]): T = lv.value
 }
 
+trait Strict[+T] extends Serializable {
```

Comment:
I added `Strict` and tend to use it when possible usually compiles a bit faster than with `Lazy`. I tend to incriminate the anonymous function (hence an additional `.class` file) that has to be created when calling the `apply` method of the `Lazy` companion. (In the v0 of my benchmark, https://github.com/alexarchambault/upickle-pprint/commit/2a63bef3a822dba4debf9726d7c515172698254f with https://github.com/alexarchambault/shapeless/commit/0fa6a9d4de45e656624668be3a8a8993df38b281, approx. ~12000 classes were generated, partly because of that, compared to less than 3000 when using macros.)

The thing with cycles is that they only need to be broken once! E.g., these typically appear in cases like
```scala
case class CC(l: List[CC])
```
or
```scala
sealed trait Tree
case class Leaf(value: Int) extends Tree
case class Node(children: List[Tree]) extends Tree
```

If one looks for a `TC[CC]`, the returned `Lazy[TC[CC]]` should look like
```scala
class anon {
  lazy val impl4: TC[HNil] = hnilTC
  lazy val impl3: TC[List[CC]] = listTC(impl1) // cycles on impl1
  lazy val impl2: TC[List[CC] :: HNil] = hconsTC(Lazy(impl3), Lazy(impl4))
  lazy val impl1: TC[CC] = genericTC(
    Generic[CC], // actually output of the Generic.materialize macro
    Lazy(impl2)
  )
}
new anon .impl1
```

So the cycle goes like
* `TC[CC]`, returned by
* `genericTC`, needs a
* `TC[List[CC] :: HNil]`, returned by
* `hconsTC`, needs a
* `TC[List[CC]]`, returned by
* `listTC`, needs a
* `TC[CC]`, hence a cycle.
Only one of the implicit instances of the involved types needs to be wrapped in a `Lazy` to break the cycle, the others can just be wrapped in a `Strict` (or not wrapped at all if implicit search doesn't diverge).